binderhub/static/js/index.js

```diff
@@ -59,50 +60,32 @@ function v2url(providerPrefix, repository, ref, path, pathType) {
   return url;
 }
 
+function loadConfig(callback) {
+  var req = new XMLHttpRequest();
+  req.onreadystatechange = function() {
+    if (req.readyState == 4 && req.status == 200)
+      callback(req.responseText)
+  };
+  req.open('GET', BASE_URL + "_config", true);
+  req.send(null);
+}
+
 function updateRepoText() {
-  var text;
-  var provider = $("#provider_prefix").val();
-  var tag_text = "Git branch, tag, or commit";
-  // first enable branch/ref field, some providers later disable it
-  $("#ref").prop("disabled", false);
-  $("label[for=ref]").prop("disabled", false);
-  if (provider === "gh") {
-    text = "GitHub repository name or URL";
-  } else if (provider === "gl") {
-    text = "GitLab.com repository or URL";
-  }
-  else if (provider === "gist") {
-    text = "Gist ID (username/gistId) or URL";
-    tag_text = "Git commit SHA";
-  }
-  else if (provider === "git") {
-    text = "Arbitrary git repository URL (http://git.example.com/repo)";
-    tag_text = "Git branch, tag, or commit SHA";
-  }
-  else if (provider === "zenodo") {
-    text = "Zenodo DOI (10.5281/zenodo.3242074)";
-    $("#ref").prop("disabled", true);
-    $("label[for=ref]").prop("disabled", true);
-  }
-  else if (provider === "figshare") {
-    text = "Figshare DOI (10.6084/m9.figshare.9782777.v1)";
-    $("#ref").prop("disabled", true);
-    $("label[for=ref]").prop("disabled", true);
-  }
-  else if (provider === "hydroshare") {
-    text = "Hydroshare resource id or URL";
-    $("#ref").prop("disabled", true);
-    $("label[for=ref]").prop("disabled", true);
-  }
-  else if (provider === "dataverse") {
-    text = "Dataverse DOI (10.7910/DVN/TJCLKP)";
-    $("#ref").prop("disabled", true);
-    $("label[for=ref]").prop("disabled", true);
+  if (Object.keys(config_dict).length === 0){
+    loadConfig(function(res) {
+      config_dict = JSON.parse(res);
```

Comment:
This is good but there is still a problem when page is loaded first time, I still see the default labels from html template:
![image](https://user-images.githubusercontent.com/6059799/75760679-f39da200-5d37-11ea-9b6e-d4a8f13eca3f.png)

This should be fixed if we update labels after this line too. We could put
```javascript
	var provider = $("#provider_prefix").val();
	var text = config_dict[provider]["text"]
	var tag_text = config_dict[provider]["tag_text"]
	var ref_prop_disabled = config_dict[provider]["ref_prop_disabled"]
	var label_prop_disabled = config_dict[provider]["label_prop_disabled"]

	$("#ref").attr('placeholder', tag_text).prop("disabled", ref_prop_disabled);
	$("label[for=ref]").text(tag_text).prop("disabled", label_prop_disabled);
	$("#repository").attr('placeholder', text);
	$("label[for=repository]").text(text);
```
into another function. And call it in the callback, after `config_dict = JSON.parse(res);` and also after if statement.

Reply:
I also noticed that the url doesn't get updated when the page first loads either. Only after explicitly selecting a provider:
![2](https://user-images.githubusercontent.com/7579677/75763908-b2ac8a00-5d45-11ea-8a87-a4d511188d11.gif)

Reply:
When `updateRepoText` is called first time, js fails because `config_dict` is empty. My suggestion should solve this problem too.

Reply:
@bitnik, I just realized that I don't think it can be done a separate PR because a repo provider class doesn't have the needed attributes, only the name. Should I make the change here instead?

Reply:
I guess you replied to wrong place? :) I think that PR can be opened when this one is done. I am fine with adding those changes also into this PR, but it would make it harder for others to review this PR.

Reply:
Yup, wrong place, sorry. For the sake of having a continuation, I'm replying here still :)
I agree it will make it harder to review, but without those changes, this PR doesn't do what it should :(

Reply:
did try this (https://github.com/jupyterhub/binderhub/pull/1038#discussion_r386889545) out?

Reply:
Yup, I tried it and it solves the problem with the default labels from the html template, but not the issue with the url :(

The gif above is with that change.

Reply:
> And call it in the callback, after config_dict = JSON.parse(res); and also after if statement.

Sorry, actually we should call in the callback in if statement and then also in else statement.